006-improving-configurability-of-kafka-listeners.md

```diff
@@ -0,0 +1,240 @@
+* Port numbers are assigned and are not configurable.
+
+Some of the use cases raised by out users in the past are listed here:
+
+### Multiple external listeners
+
+In some situations, having multiple external listener could be very useful.
+Apart from the obvious such as multiple authentication mechanisms or listeners with and without encryption, multiple listeners can be also useful to handle access from different networks outside the Kubernetes cluster.
+
+For example in AWS, you can have two types of load balancers: public and internal.
+Public load balancers are exposed tot he Internet and can be used by applications running anywhere online.
+On the other hand internal load balancers are exposed only within the VPC (Virtual Private Cloud) - and accessible only to the applications running outside of Kubernetes but inside the private network of the user.
+
+### Configurability of internal listeners
+
+Some users heavily customize their Kubernetes network.
+One of the examples is joining their Kubernetes network with their network outside of Kubernetes.
+In such case, a common problem is using or not using the cluster service DNS suffix (`.cluster.local` by default) which is in some cases needed and in some not.
+Having the ability to override the advertised hostnames on internal listeners similarly to how they can be treated on the external listener or being able to have ore internal listeners could help in these situations.
+
+## Proposed changes
+
+This proposal tries to make the configuration of the listeners more flexible.
+It should use array / list instead of an object with predefined listeners.
+It should also open more configuration options to all of them.
+
+_Note: The replication listener should be still kept non-configurable and hidden from users._
+
+The new configuration can look like this:
+
+```yaml
+    listeners:
+      - name: encrypted1
+        port: 9092
+        type: service
+        tls: true
+        authentication:
+          type: tls
+        overrides:
+          brokers:
+            - broker: 0
+                advertisedHost: my-cluster-brokers-kafka-0.myns.cluster.local
+            - broker: 1
+                advertisedHost: my-cluster-brokers-kafka-1.myns.cluster.local
+            - broker: 2
+                advertisedHost: my-cluster-brokers-kafka-2.myns.cluster.local
+      - name: encrypted2
+        port: 9093
+        type: service
+        tls: true
+        authentication:
+          type: acram-sha-512
+      - name: routes
+        port: 9094
+        type: route
+        tls: true
+        authentication:
+          type: tls
+      - name: nodeports
+        port: 9095
+        type: nodeport
+        tls: true
+        authentication:
+          type: tls
+```
+
+The listeners will be configurable as array.
+Each listener will have its own unique name and will specify the port and type as required values.
```

Comment:
```suggestion
Each listener will have its own unique name and will specify a type and unique port as required values.
```

Since we are banning 9091 and 9094, I don't think it will be much extra work to iterate through the defined listeners and ensure there are no duplicate port usages.